Bounds as inputs and chips outputs

crangelsmith opened the pull request:
Fixes #8 and #25

- [ ] Move from 'coords' to 'bounds' in the pipeline input and main attribute of the Downloader and Procesor classes (bounds are list with 4 entries as left, bottom, right, top).
- [ ] Remove 'region_size' option as not used anymore
- [ ] After the tiff files are downloaded the bounds are read from the tiff files and used downstream (given a projection, this is now in eastings, northings instead of lat long).
- [ ] Processed images (chips and RGB images use the updated projected bounds in their naming convention).
- [ ] Added the date range in the output names, in order to provide all relevant info in the name.
- [ ] Chips are now identified with the left_bottom corner + chip dimension + date range used in the download. 
- [ ] Count data is now stored at the chip level as a json file with summary statistics (mean, std, min,max,etc), it has same name as the chip array.
- [ ] Removed config files that do not follow the new nomenclature of bounds instead of coords.

crangelsmith commented:
This is how the new output of the pipeline looks like:

<img width="868" alt="Screenshot 2022-08-11 at 14 11 37" src="https://user-images.githubusercontent.com/11162074/184141272-32ea1cc1-f942-4e35-aee9-f29e98f2799e.png">


Notice that the directory with the highest level still has the lat,long bounds, but all processed images are now in eastings and northings.

andrewphilipsmith commented on pyveg/src/download_modules.py:
```diff
@@ -126,11 +125,11 @@ def prep_data(self, date_range):
 
         image_coll = ee.ImageCollection(self.collection_name)
-        geom = ee.Geometry.Point(self.coords)
+        geom = ee.Geometry.Point(get_coords(self.bounds))
```
We can specify the coordinate reference system in this line (https://developers.google.com/earth-engine/apidocs/ee-geometry-point).

I _think_ if we do that and keep everything in BNG then we might get more consistent download image dimensions.

crangelsmith replied:
It worked :)

andrewphilipsmith commented on pyveg/src/download_modules.py:
```diff
@@ -126,11 +127,25 @@ def prep_data(self, date_range):
         url_list:  a list of URLs from which zipfiles can be downloaded from GEE.
         """
-        region = get_region_string(self.coords, self.region_size)
+        region = get_region_string(self.bounds)
```
I know I wrote this (!) but:

1) Need to remove all the debug statements (minor)
2) Need to decide whether to go with hardcoding BNG (EPSG:27700) throughout. Or make this a parameter defined in the config. If the projection is not specified, it defaults to WGS84 Long/Lat
https://developers.google.com/earth-engine/apidocs/ee-geometry
3) Note that the `region` variable is not used anymore.

crangelsmith replied:
Debug statements and region now removed.

I've made the projection now configurable and defaulting to BNG if not configuration is given. The issue I see atm is that a lot of the name formatting now is assuming BNG, and needs to be adapted if is not (it can look weird in lat long). I think options for this can be added later on once we are tyding out the code given that we only care about BNG for this project.

andrewphilipsmith commented on pyveg/tests/test_download_modules.py:
```diff
@@ -52,7 +52,7 @@ def test_veg_downloader_configure_sentinel2():
     veg_downloader.RGB_bands = ["B4", "B3", "B2"]
     veg_downloader.NIR_band = "B8"
-    veg_downloader.coords = (11.45, 27.5)
+    veg_downloader.bounds = [-3.0183, 53.3649, -2.9482, 53.4350]
```
Needs some tests using BNG values rather than Long/Lat

crangelsmith replied:
fixed now.

andrewphilipsmith commented on pyveg/tests/test_pipeline_core.py:
```diff
@@ -29,11 +29,11 @@ def test_add_sequence_to_pipeline():
     s = Sequence("testseq")
     assert s.is_configured == False
-    s.coords = [1.23, 4.56]
+    s.bounds = [-3.01, 53.36, -2.94, 53.43]
```
Needs some tests using BNG values rather than Long/Lat

crangelsmith replied:
fixed.

andrewphilipsmith commented on pyveg/tests/test_processor_modules.py:
```diff
@@ -36,13 +36,13 @@ def test_Sentinel2_image_processor():
     vip.output_location = tmp_png_path
     vip.ndvi = True
-    vip.coords = [11.58, 27.95]
+    vip.bounds = [-3.0183, 53.3649, -2.9482, 53.4350]
```
Needs some tests using BNG values rather than Long/Lat

crangelsmith replied:
fixed now.

andrewphilipsmith commented on pyveg/src/coordinate_utils.py:
```diff
@@ -8,10 +8,29 @@
 
 
-def get_region_string(coords, region_size):
+def get_coords(bounds):
```
I don't think that this function is required any more.

crangelsmith replied:
I'm leaving these functions there for now in the utils, just in case we want to use them later. But we should remove them once the pipeline is fixed.

andrewphilipsmith reviewed (commented):
As discussed:

* The bits I've added need tidying up 😱 
* Before splitting into chips, confirm that the downloaded image is 1024*1024 and error if not.
* There are some unused imports and functions. These don't have to be dealt with in this PR, but can be identified with: 
```
$ pylint pyveg/src/* | grep W0611
```

* (Plan B) If we still get lots of downloaded images that are not 1024*1024, then we could try using `rasterio clip` to the downloaded images to real-world coordinates, rather than pixel coordinates.

crangelsmith commented on pyveg/src/coordinate_utils.py:
```diff
@@ -8,10 +8,29 @@
+    return [x, y]
+
+
+def get_region_string(bounds):
```
Same as the comment below, leaving these functions there for now in the utils, just in case we want to use them later. But we should remove them once the pipeline is fixed.

crangelsmith commented:
> As discussed:
> 
> * The bits I've added need tidying up 😱
> * Before splitting into chips, confirm that the downloaded image is 1024*1024 and error if not.
> * There are some unused imports and functions. These don't have to be dealt with in this PR, but can be identified with:
> 
> ```
> $ pylint pyveg/src/* | grep W0611
> ```
> 
> * (Plan B) If we still get lots of downloaded images that are not 1024*1024, then we could try using `rasterio clip` to the downloaded images to real-world coordinates, rather than pixel coordinates.

I've added a fail and logging error if the downloaded images do not have the right size or if the bounds are not the same as the input. It seems to be ok for the cases I have tested. 

Also, I've checked that the chips created in the pipeline map to the geoparquet file (using chip_id) and that seems to be working well too :-)
